### PR TITLE
Optimized the AND and OR FEEL operations 

### PR DESCRIPTION
**Pull request by SimonRinguette**
to not evaluate the right operand if not required

### CODE REVIEW

**Comment by SimonRinguette** [comment]
@tarilabs As discussed by email

**Comment by tarilabs** [comment]
ok to test

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java [hunk 1]
```diff
@@ -163,41 +163,63 @@ public Type getResultType() {
...
+                Boolean leftAND = EvalHelper.getBooleanOrNull(this.left.evaluate(ctx));
+                if (leftAND != null) {
+                    if (leftAND.booleanValue()) {
+                        return EvalHelper.getBooleanOrNull(this.right.evaluate(ctx));
+                    } else {
+                        return Boolean.FALSE; //left hand operand is false, we do not need to evaluate right side
+                    }
+                } else {
+                    Boolean rightAND = EvalHelper.getBooleanOrNull(this.right.evaluate(ctx));
+                    return rightAND == Boolean.FALSE ? rightAND : null;
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return rightAND == Boolean.FALSE ? false : null;
```

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java [hunk 2]
```diff
@@ -163,41 +163,63 @@ public Type getResultType() {
...
+                Boolean leftOR = EvalHelper.getBooleanOrNull(this.left.evaluate(ctx));
+                if (leftOR != null) {
+                    if (!leftOR.booleanValue()) {
+                        return EvalHelper.getBooleanOrNull(this.right.evaluate(ctx));
+                    } else {
+                        return Boolean.TRUE; //left hand operand is true, we do not need to evaluate right side
+                    }
+                } else {
+                    Boolean rightOR = EvalHelper.getBooleanOrNull(this.right.evaluate(ctx));
+                    return rightOR == Boolean.TRUE ? rightOR : null;
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return rightOR == Boolean.TRUE ? true : null;
```

**Review by tarilabs — APPROVED** [review]
I noted a couple of line comments, mainly I'm curious to see if SonarCloud will highlight them as I noted, or not.

PR is fine for me as long as it will be green

Thanks again Simon!

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java (hunk 1 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return rightAND.equals(Boolean.FALSE) ? false : null;
```

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java (hunk 2 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return rightOR.equals(Boolean.TRUE) ? true : null;
```

**Review by tarilabs — CHANGES_REQUESTED** [review]
(updated)

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java (hunk 1 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return Boolean.FALSE.equals(rightAND) ? Boolean.FALSE : null;
```

File: kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java (hunk 2 above)
Comment (tarilabs) [reply kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/lang/ast/InfixOpNode.java]:
```suggestion
                    return Boolean.TRUE.equals(rightOR) ? Boolean.TRUE : null;
```

**Review by tarilabs — CHANGES_REQUESTED** [review]
@SimonRinguette can you kindly implement these last couple of code change please, it should be possible to accept them directly from this GH page, since the .equals is needed to improve stats of static code analysis which otherwise is bad marked.

Ref: https://sonarcloud.io/project/issues?id=org.drools%3Adrools&open=AXOgIgypijryR0MKiFIh&pullRequest=3016&resolved=false&types=BUG

thanks!

**Comment by SimonRinguette** [comment]
I've accepted your suggestions, make sure to SQUASH those 2 extra commits

**Comment by tarilabs** [comment]
> make sure to SQUASH those 2 extra commits

affirmative, I always request gatekeeper to squash on merge ;) and that should also be their default

**Review by hellowdan — APPROVED** [review]

**Comment by tarilabs** [comment]
can a @kiegroup/gatekeepers kindly **merge with squash**, pls?

**Comment by mareknovotny** [comment]
so I guess you approved this PR @tarilabs, right?

**Comment by tarilabs** [comment]
Yes as can be seen here https://github.com/kiegroup/drools/pull/3016#commits-pushed-53a3c46 the requested code changes were applied.

**Comment by mareknovotny** [comment]
@tarilabs well, I usually look at status on top right corner where you requested changes ;) reading all comments is the second step, but if you don't Approve it is ambiguous.

**Review by tarilabs — APPROVED** [review]
Minor code changes requested with https://github.com/kiegroup/drools/pull/3016#pullrequestreview-458969283
were applied by OP with https://github.com/kiegroup/drools/pull/3016#commits-pushed-53a3c46

**Comment by tarilabs** [comment]
can a @kiegroup/gatekeepers kindly merge with squash, pls? thx in advance

**Pull request merged**